proxylib/cassandra/cassandraparser.go

```diff
@@ -0,0 +1,643 @@
+			} else if res == actionNoTable && cr.tableRegexCompiled != nil {
+				ParseError(fmt.Sprintf("query_action '%s' is not compatible with a query_table match", cr.queryActionExact), rule)
+			}
+
+		}
+
+		log.Debugf("Parsed CassandraRule pair: %v", cr)
+		rules = append(rules, &cr)
+	}
+	return rules
+}
+
+type CassandraParserFactory struct{}
+
+var cassandraParserFactory *CassandraParserFactory
+
+func init() {
+	log.Info("init(): Registering cassandraParserFactory")
+	RegisterParserFactory("cassandra", cassandraParserFactory)
+	RegisterL7RuleParser("cassandra", CassandraRuleParser)
+}
+
+type CassandraParser struct {
+	connection *Connection
+	inserted   bool
+	keyspace   string // stores current keyspace name from 'use' command
+
+	// stores prepared query string while
+	// waiting for 'prepared' reply from server
+	// with a prepared id.
+	// replies associated via stream-id
+	preparedQueryPathByStreamID map[uint16]string
+
+	// allowing us to enforce policy on query
+	// at the time of the execute command.
+	preparedQueryPathByPreparedID map[string]string // stores query string based on prepared-id,
+}
+
+func (pf *CassandraParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("CassandraParserFactory: Create: %v", connection)
+
+	p := CassandraParser{connection: connection}
+	p.preparedQueryPathByStreamID = make(map[uint16]string)
+	p.preparedQueryPathByPreparedID = make(map[string]string)
+	return &p
+}
+
+func (p *CassandraParser) OnData(reply, endStream bool, dataArray [][]byte, offset int) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte(""))[offset:]
+	log.Debugf("OnData offset  = %d", offset)
+
+	if len(data) < cassHdrLen {
+		// Partial header received, ask for more
+		needs := cassHdrLen - len(data)
+		log.Debugf("Did not receive full header, need %d more bytes", needs)
+		return MORE, needs
+	}
+
+	// full header available, read full request length
+	requestLen := binary.BigEndian.Uint32(data[5:9])
+	log.Debugf("Request length = %d", requestLen)
+	if requestLen > cassMaxLen {
+		log.Errorf("Request length of %d is greater than 256 MB", requestLen)
+		return ERROR, 0
+	}
+
+	dataMissing := (cassHdrLen + int(requestLen)) - len(data)
+	if dataMissing > 0 {
+		// full header received, but only partial request
+
+		log.Debugf("Hdr received, but need %d more bytes of request", dataMissing)
+		return MORE, dataMissing
+	}
+
+	// we don't parse reply traffic for now
+	if reply {
+		if len(data) == 0 {
+			log.Debugf("ignoring zero length reply call to onData")
+			return NOP, 0
+
+		}
+		cassandraParseReply(p, data[0:(cassHdrLen+requestLen)])
+
+		log.Debugf("reply, passing %d bytes", uint32(len(data)))
+		return PASS, len(data)
+	}
+
+	err, paths := cassandraParseRequest(p, data[0:(cassHdrLen+requestLen)])
+	if err != 0 {
+		log.Errorf("Parsing error %d", err)
+		return ERROR, 0
```

Comment:
As `err` is of type `OpError`, you should return it like so:
```
	return ERROR, int(err)
```

Reply:
Could also return error detail in `paths` and access log it as DROP, passing the error detail in "status" field in L7LogEntry. "status" is recognized by cilium monitor and shown in monitor output. That would make parse errors more visible.